rcl/src/rcl/security.c

```diff
@@ -0,0 +1,351 @@
+    if (-1 == tinydir_readfile(&dir, &file)) {
+      goto cleanup;
+    }
+    if (file.is_dir) {
+      size_t matched_name_length = strnlen(file.name, sizeof(file.name) - 1);
+      if (0 ==
+        strncmp(
+          file.name, name, matched_name_length) &&
+        matched_name_length > max_match_length)
+      {
+        max_match_length = matched_name_length;
+        memcpy(matched_name, file.name, max_match_length);
+      }
+    }
+    if (-1 == tinydir_next(&dir)) {
+      goto cleanup;
+    }
+  }
+cleanup:
+  tinydir_close(&dir);
+  return max_match_length > 0;
+}
+
+char * exact_match_lookup(
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator)
+{
+  // Perform an exact match for the node/context's name in directory <root dir>/<namespace>.
+  char * secure_root = NULL;
+  // "/" case when root namespace is explicitly passed in
+  if (1 == strlen(namespace_)) {
+    secure_root = rcutils_join_path(ros_secure_root_env, name, *allocator);
+  } else {
+    char * fqn = NULL;
+    char * root_path = NULL;
+    // Combine namespace with name
+    // TODO(ros2team): remove the hard-coded value of the root namespace
+    fqn = rcutils_format_string(*allocator, "%s%s%s", namespace_, "/", name);
+    // Get native path, ignore the leading forward slash
+    // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead
+    root_path = rcutils_to_native_path(fqn + 1, *allocator);
+    secure_root = rcutils_join_path(ros_secure_root_env, root_path, *allocator);
+    allocator->deallocate(fqn, allocator->state);
+    allocator->deallocate(root_path, allocator->state);
+  }
+  return secure_root;
+}
+
+char * prefix_match_lookup(
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator)
+{
+  // Perform longest prefix match for the node/context's name in directory <root dir>/<namespace>.
+  char * secure_root = NULL;
+  char matched_dir[_TINYDIR_FILENAME_MAX] = {0};
+  char * base_lookup_dir = NULL;
+  if (strlen(namespace_) == 1) {
+    base_lookup_dir = (char *) ros_secure_root_env;
+  } else {
+    // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead.
+    base_lookup_dir = rcutils_join_path(ros_secure_root_env, namespace_ + 1, *allocator);
+  }
+  if (get_best_matching_directory(base_lookup_dir, name, matched_dir)) {
+    secure_root = rcutils_join_path(base_lookup_dir, matched_dir, *allocator);
+  }
+  if (base_lookup_dir != ros_secure_root_env && NULL != base_lookup_dir) {
+    allocator->deallocate(base_lookup_dir, allocator->state);
+  }
+  return secure_root;
+}
+
+char * rcl_get_secure_root(
+  const char * name,
+  const char * namespace_,
+  const rcl_allocator_t * allocator)
+{
+  bool ros_secure_directory_override = true;
+  bool use_node_name_in_lookup = rmw_use_node_name_in_security_directory_lookup();
```

Comment:
It appears that this is not used, is that true?

Reply:
Fixed in https://github.com/ros2/rcl/pull/515/commits/533ea2bbd7a56468cedca1323de9f5f176218920.